feat: Support snap to pixel for RectangleLabels

niklub opened the pull request:
Added ability to snap bounding box annotation to the pixel grid via:
```xml
<RectangleLabels name="bbox" toName="image" snap="pixel"/>
```

https://www.loom.com/share/a362e1747c2543e487be344ca20f4827?sid=05df6648-222e-4f80-8ede-64adc71a195f
<!--

This description MUST be filled out for a PR to receive a review. Its primary purposes are:

 - to enable your reviewer to review your code easily, and
 - to convince your reviewer that your code works as intended.

Some pointers to think about when filling out your PR description:
 - Reason for change: Description of problem and solution
 - Screenshots: All visible changes should include screenshots.
 - Rollout strategy: How will this code be rolled out? Feature flags / env var / other
 - Testing: Description of how this is being verified
 - Risks: Are there any known risks associated with this change, eg to security or performance?
 - Reviewer notes: Any info to help reviewers approve the PR
 - General notes: Any info to help onlookers understand the code, or callouts to significant portions.

You may use AI tools such as Copilot Actions to assist with writing your PR description (see https://docs.github.com/en/copilot/using-github-copilot/using-github-copilot-for-pull-requests/creating-a-pull-request-summary-with-github-copilot); however, an AI summary isn't enough by itself. You'll need to provide your reviewer with strong evidence that your code works as intended, which requires actually running the code and showing that it works.

-->

Commit:
feat: Support snap to pixel for RectangleLabels

robot-ci-heartex committed:
ci: Build tag docs

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16297883875

hlomzik commented on web/libs/editor/src/regions/RectRegion.jsx:
```diff
@@ -492,7 +578,46 @@ const HtxRectangleView = ({ item, setShapeRef }) => {
+          rotation,
+        );
+      } else {
+        item.setPosition(x, y, width, height, rotation);
```
`setPosition()` already has snapping logic, so this part can be left as it was before

hlomzik commented on web/libs/editor/src/regions/RectRegion.jsx:
```diff
@@ -467,13 +521,45 @@ const HtxRectangleView = ({ item, setShapeRef }) => {
+          rotation,
+        );
+      } else {
+        item.setPosition(x, y, width, height, rotation);
```
The same here — logic is already in `setPosition()`

Commit:
Address comments, remove duplicated coe

robot-ci-heartex committed:
ci: Build tag docs

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16318818711

Commit:
Remove unecesseray changes

Commit:
more updates

hlomzik commented:
We should test regions restoration via:
- load submitted annotation
- switch between annotations
- undo/redo
- history switching
- switching between images for MIG

Code looks good! Would be also nice to have screenshots/videos in PR to see it in action.

Commit:
Fix rerender issue

Commit:
snap before commit drawing

hlomzik commented on web/libs/editor/src/regions/RectRegion.jsx:
```diff
@@ -478,6 +498,14 @@ const HtxRectangleView = ({ item, setShapeRef }) => {
       t.setAttr("scaleY", 1);
 
+      // Sync Konva shape attributes back to computed canvas coordinates to cause a re-render
+      t.position({
```
better to wrap this with `if (self.control?.snap === "pixel")` to avoid unnecessary re-renders + add a comment that these canvas values are updated inside `setPosition()`
+apply this to the same code in `onDragEnd`

Commit:
Add test (wip)

Commit:
address comments

Commit:
ensure proper scaling for 3click bbox

Commit:
fmt

Commit:
Remove test file

hlomzik commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/16422955594)
> Successfully merged:  create mode 100644 web/libs/editor/tests/integration/e2e/video/timeline_region_loop.cy.ts

hlomzik committed:
Merge branch 'develop' into 'fb-bros-187'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/16422955594
